### PR TITLE
feat(instrumentation): integrate analytics report workflow CLI-302

### PR DESCRIPTION
**Pull request by mgyorke**
## Pull Request Submission

Please check the boxes once done.

The pull request must:

- **Reviewer Documentation**
    - [ ] follow [CONTRIBUTING](https://github.com/snyk/cli/blob/main/CONTRIBUTING.md) rules
    - [ ] be accompanied by a detailed description of the changes
    - [ ] contain a risk assessment of the change (Low | Medium | High) with regards to breaking existing functionality. A change e.g. of an underlying language plugin can completely break the functionality for that language, but appearing as only a version change in the dependencies.
    - [ ] highlight breaking API if applicable
    - [ ] contain a link to the automatic tests that cover the updated functionality.
    - [ ] contain testing instructions in case that the reviewer wants to manual verify as well, to add to the manual testing done by the author.
    - [ ] link to the link to the PR for the User-facing documentation
- **User facing Documentation**
    - [ ] update any relevant documentation in gitbook by submitting a gitbook PR, and including the PR link here
    - [ ] ensure that the message of the final single commit is descriptive and prefixed with either `feat:` or `fix:` , others might be used in rare occasions as well, if there is no need to document the changes in the release notes. The changes or fixes should be described in detail in the commit message for the changelog & release notes.
- **Testing**
    - [ ] Changes, removals and additions to functionality must be covered by acceptance / integration tests or smoke tests - either already existing ones, or new ones, created by the author of the PR.

## Pull Request Review

All pull requests must undergo a thorough review process before being merged. 
The review process of the code PR should include code review, testing, and any necessary feedback or revisions. 
Pull request reviews of functionality developed in other teams only review the given documentation and test reports. 

Manual testing will not be performed by the reviewing team, and is the responsibility of the author of the PR.

For Node projects: It’s important to make sure changes in `package.json` are also affecting `package-lock.json` correctly.

****************************If a dependency is not necessary, don’t add it.**************************** 

When adding a new package as a dependency, make sure that the change is absolutely necessary. We would like to refrain from adding new dependencies when possible.
Documentation PRs in gitbook are reviewed by Snyk's content team. They will also advise on the best phrasing and structuring if needed.

## Pull Request Approval

Once a pull request has been reviewed and all necessary revisions have been made, it is approved for merging into 
the main codebase. The merging of the code PR is performed by the code owners, the merging of the documentation PR 
by our content writers.

## Any background context you want to provide?
This PR depends on [this](https://github.com/snyk/go-application-framework/pull/184) and [this](https://github.com/snyk/go-application-framework/pull/191) pull requests.

### CODE REVIEW

File: cliv2/cmd/cliv2/main.go [hunk 1]
```diff
@@ -242,15 +241,41 @@ func sendAnalytics(analytics analytics.Analytics, debugLogger *zerolog.Logger) {
 	}
 }
 
-func sendInstrumentation(instrumentor analytics.InstrumentationCollector, logger *zerolog.Logger) {
-	// TODO: actually send data once CLI-303 is implemented
+func sendInstrumentation(eng workflow.Engine, instrumentor analytics.InstrumentationCollector, logger *zerolog.Logger) {
+	// Avoid duplicate data to be send for IDE integrations that use the CLI
+	if utils.IsSnykIde(os.Getenv("SNYK_INTEGRATION_NAME")) {
```
Comment (PeterSchafer) [inline cliv2/cmd/cliv2/main.go:246]: Suggestion: Please use the configuration to access values, see [here](https://github.com/snyk/go-application-framework/blob/main/pkg/local_workflows/auth_workflow.go#L84). This way, there is no need for the implementation to know where the values come from and how they exactly need to be named.

Comment (mgyorke) [reply cliv2/cmd/cliv2/main.go]: fixed!

File: cliv2/cmd/cliv2/main.go [hunk 2]
```diff
@@ -14,35 +12,36 @@ import (
...
 	"github.com/snyk/cli-extension-sbom/pkg/sbom"
+	"github.com/snyk/cli/cliv2/internal/cliv2"
+	"github.com/snyk/cli/cliv2/internal/constants"
 	"github.com/snyk/container-cli/pkg/container"
 	"github.com/snyk/go-application-framework/pkg/analytics"
 	"github.com/snyk/go-application-framework/pkg/app"
 	"github.com/snyk/go-application-framework/pkg/auth"
 	"github.com/snyk/go-application-framework/pkg/configuration"
+	"github.com/snyk/go-application-framework/pkg/instrumentation"
+	_ "github.com/snyk/go-application-framework/pkg/networking/fips_enable"
```
Comment (PeterSchafer) [inline cliv2/cmd/cliv2/main.go:28]: Issue: This import needs to be on top, please see the related comment.

File: cliv2/cmd/cliv2/main.go (hunk 1 above)
Comment (PeterSchafer) [reply cliv2/cmd/cliv2/main.go]: unfortunately not, please take a closer look at the linked code snippet.

File: cliv2/cmd/cliv2/main.go (hunk 2 above)
Comment (mgyorke) [reply cliv2/cmd/cliv2/main.go]: good catch, the gopls moves these imports on auto format. Pushed fix.

File: cliv2/cmd/cliv2/main.go (hunk 1 above)
Comment (mgyorke) [reply cliv2/cmd/cliv2/main.go]: fixed the fix.

File: cliv2/cmd/cliv2/main.go [hunk 3]
```diff
@@ -242,15 +241,42 @@ func sendAnalytics(analytics analytics.Analytics, debugLogger *zerolog.Logger) {
...
+		"application/json",
+		v2InstrumentationData,
+	)
+
+	logger.Trace().Msg("Invoking report analytics workflow")
+
+	_, err = eng.InvokeWithInputAndConfig(
+		localworkflows.WORKFLOWID_REPORT_ANALYTICS,
+		[]workflow.Data{inputData},
+		globalConfiguration,
```
Comment (PeterSchafer) [inline cliv2/cmd/cliv2/main.go:272]: Issue: Did you test the implementation? when I run it, I see an error like the one below
> 2024-05-28T09:30:25Z main - Failed to invoke analytics error="set `--experimental` flag to enable analytics report command"

when I add the flag to the command line the workflow runs through. But we can't require a=our users to set the `experimental` flag. this needs to be done programmatically here.

File: cliv2/cmd/cliv2/main.go [hunk 4]
```diff
@@ -242,15 +242,46 @@ func sendAnalytics(analytics analytics.Analytics, debugLogger *zerolog.Logger) {
...
 	v2InstrumentationData := utils.ValueOf(json.Marshal(data))
 	logger.Trace().Msgf("Instrumentation: %v", string(v2InstrumentationData))
+
+	inputData := workflow.NewData(
+		workflow.NewTypeIdentifier(localworkflows.WORKFLOWID_REPORT_ANALYTICS, "reportAnalytics"),
+		"application/json",
+		v2InstrumentationData,
+	)
+
+	logger.Trace().Msg("Invoking report analytics workflow")
```
Comment (PeterSchafer) [inline cliv2/cmd/cliv2/main.go:268]: Suggestion: let's remove the workflow name in logs and comments, the name is not perfect, so let's not spread it more than necessary.

File: cliv2/cmd/cliv2/main.go [hunk 5]
```diff
@@ -242,15 +242,46 @@ func sendAnalytics(analytics analytics.Analytics, debugLogger *zerolog.Logger) {
...
+	// we pass the flag here so that we report at every interaction
+	localConfiguration.Set(configuration.FLAG_EXPERIMENTAL, true)
+	_, err = eng.InvokeWithInputAndConfig(
+		localworkflows.WORKFLOWID_REPORT_ANALYTICS,
+		[]workflow.Data{inputData},
+		localConfiguration,
+	)
+
+	if err != nil {
+		logger.Err(err).Msg("Failed to invoke analytics")
```
Comment (PeterSchafer) [inline cliv2/cmd/cliv2/main.go:281]: Suggestion: Same as above! Also success and failure logs should use the same wording.

File: test/jest/acceptance/snyk-test/all-projects.spec.ts [hunk 6]
```diff
@@ -145,9 +145,11 @@ describe('snyk test --all-projects (mocked server only)', () => {
     );
 
     expect(code).toEqual(1);
-    const req = server.popRequest();
-    expect(req.query.ignorePolicy).toBeTruthy(); // should request to ignore the policy
+    const requests = server.getRequests().filter((req: any) => {
+      if (req.query.ignorePolicy) return req.query.ignorePolicy;
```
Comment (PeterSchafer) [inline test/jest/acceptance/snyk-test/all-projects.spec.ts:149]: Suggestion: The statement in `filter()` should return a truthy or falsy value (see for example [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter#callbackfn)). This means that you can make all the statements simpler.

```suggestion
      return (req.query.ignorePolicy !== undefined);
```

File: test/jest/acceptance/snyk-test/all-projects.spec.ts [hunk 7]
```diff
@@ -169,8 +171,16 @@ describe('snyk test --all-projects (mocked server only)', () => {
...
+    const backendRequests = server.getRequests().filter((req: any) => {
+      if (
+        req.method === 'POST' &&
+        req.headers['x-snyk-cli-version'] &&
+        req.url.includes('/api/v1/test-dep-graph')
+      )
+        return req;
+    });
+
+    expect(backendRequests.length).toBeGreaterThan(0);
```
Comment (PeterSchafer) [inline test/jest/acceptance/snyk-test/all-projects.spec.ts:183]: Issue: This  changes the expectation, `> 0` is not the same as `==2`.  You should be able to filter the requests and leave the expectations unchanged.

File: test/jest/acceptance/snyk-test/all-projects.spec.ts [hunk 8]
```diff
@@ -169,8 +171,16 @@ describe('snyk test --all-projects (mocked server only)', () => {
         ? 'vulnerable\\package-lock.json'
         : 'vulnerable/package-lock.json';
 
-    const backendRequests = server.popRequests(2);
-    expect(backendRequests).toHaveLength(2);
+    const backendRequests = server.getRequests().filter((req: any) => {
```
Comment (PeterSchafer) [inline test/jest/acceptance/snyk-test/all-projects.spec.ts:174]: Suggestion: The same as above! But here you could probably just filter for the url. The more unrelated values you expect, the higher the chance that this test breaks due to an unrelated change.

File: test/jest/acceptance/snyk-test/protect-upgrade-notification.spec.ts [hunk 9]
```diff
@@ -52,20 +52,25 @@ describe('analytics module', () => {
...
-          metadata: {
-            'upgradable-snyk-protect-paths': 1,
+    const requests = server.getRequests().filter((req) => {
+      if (JSON.stringify(req.body).includes('upgradable-snyk-protect-paths')) {
+        return req;
+      }
+    });
+
+    expect(requests.length).toBeGreaterThan(0);
+    requests.forEach((req) => {
```
Comment (PeterSchafer) [inline test/jest/acceptance/snyk-test/protect-upgrade-notification.spec.ts:62]: Question: Are these really multiple requests? Looks like before there was only one.

File: test/jest/acceptance/snyk-test/protect-upgrade-notification.spec.ts [hunk 10]
```diff
@@ -169,19 +184,27 @@ describe('analytics module', () => {
...
-
-    const lastRequest = server.popRequest();
-    expect(lastRequest).toMatchObject({
-      query: {},
-      body: {
-        data: {
-          command: 'test',
-          metadata: {
-            'upgradable-snyk-protect-paths': 0,
+    // server.popRequest();
```
Comment (PeterSchafer) [inline test/jest/acceptance/snyk-test/protect-upgrade-notification.spec.ts:187]: Suggestion: remove commented code

File: cliv2/cmd/cliv2/main.go (hunk 4 above)
Comment (mgyorke) [reply cliv2/cmd/cliv2/main.go]: You're right! I changed the logs on that line and around it, I think it's more inline with what we want.

File: test/jest/acceptance/snyk-test/all-projects.spec.ts (hunk 7 above)
Comment (mgyorke) [reply test/jest/acceptance/snyk-test/all-projects.spec.ts]: My reasoning here, and in the other tests, was that the number of requests is not that important as much as having the right requests. So, that line was an exit "trap" for when there are no requests.

But I also see the value in checking the number, so I changed it to do an exact check for each test.

File: test/jest/acceptance/snyk-test/all-projects.spec.ts (hunk 6 above)
Comment (mgyorke) [reply test/jest/acceptance/snyk-test/all-projects.spec.ts]: I refactored this and the other instances.

**Review by thisislawatts — APPROVED** [review]
